ducktape/cluster/json.py

```diff
@@ -12,29 +12,45 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from __future__ import absolute_import
+
+from ducktape.command_line.config import ConsoleConfig
 from .cluster import Cluster, ClusterSlot
 from .remoteaccount import RemoteAccount
-
 import collections, json, os, os.path
 
 
 class JsonCluster(Cluster):
     """
     An implementation of Cluster that uses static settings specified in a cluster file.
+
+    - If cluster_json is specified, use cluster info from it
+    - Otherwise
+      - If cluster_file is specified in the constructor's kwargs, read cluster info from the file specified by cluster_file
+      - Otherwise, read cluster info from the default file specified by ConsoleConfig.CLUSTER_FILE.
     """
 
-    def __init__(self, cluster_json=None):
+    def __init__(self, cluster_json=None, *args, **kwargs):
         super(JsonCluster, self).__init__()
         if cluster_json is None:
-            cluster_json_path = os.path.abspath(os.path.join(os.getcwd(), "cluster.json"))
-            cluster_json = json.load(open(cluster_json_path))
+            # This is a directly instantiation of JsonCluster rather than from a subclass (e.g. VagrantCluster)
+            cluster_file = kwargs.get("cluster_file")
+            if cluster_file is None:
+                cluster_file = ConsoleConfig.CLUSTER_FILE
+            cluster_json = json.load(open(os.path.abspath(cluster_file)))
+
         try:
-            init_nodes = [RemoteAccount(ninfo["hostname"], ninfo.get("user"), ninfo.get("ssh_args"),
-                                        ssh_hostname=ninfo.get("ssh_hostname"))
-                                        for ninfo in cluster_json["nodes"]]
+            node_accounts = [RemoteAccount(ninfo["hostname"], ninfo.get("user"), ninfo.get("ssh_args"),
+                                           ssh_hostname=ninfo.get("ssh_hostname"),
+                                           externally_routable_ip=ninfo.get("externally_routable_ip"))
+                                           for ninfo in cluster_json["nodes"]]
+            for node_account in node_accounts:
+                if node_account.externally_routable_ip is None:
+                    node_account.externally_routable_ip = self._externally_routable_ip(node_account)
```

Comment:
Can you elaborate what is the negative effect of doing this? I think this will have the same overhead as previous approach, right? During initialization, an ssh command will be executed on each remote machine before and after this patch.